### PR TITLE
swss-common LIB update for events publish/subscribe

### PR DESCRIPTION
**Pull request by renukamanavalan**

### CODE REVIEW

File: common/events_common.h [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+ * but that class needs some additional support, that declares
+ * boost::serialization::access as private friend and couple more tweaks.
+ * The std::map & vector inherently supports serialization.
+ */
+template <typename Map>
+int
+serialize(const Map& data, string &s)
+{
+    s.clear();
+    std::stringstream _ser_ss;
```
Comment (zbud-msft) [inline common/events_common.h]: Inconsistent use of std::

File: common/events_common.h [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+/*
+ * Way to serialize map or vector
+ * boost::archive::text_oarchive could be used to archive any struct/class
+ * but that class needs some additional support, that declares
+ * boost::serialization::access as private friend and couple more tweaks.
+ * The std::map & vector inherently supports serialization.
+ */
+template <typename Map>
+int
+serialize(const Map& data, string &s)
```
Comment (zbud-msft) [inline common/events_common.h:174]: Why does serialize require address of Map?

File: common/events.cpp [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+                else {
+                    break;
+                }
+            }
+            else {
+                serialize(evt_data, evt_str);
+                events.push_back(evt_str);
+            }
+            chrono::steady_clock::time_point now = chrono::steady_clock::now();
+            if (chrono::duration_cast<std::chrono::milliseconds>(now - start).count() >
```
Comment (zbud-msft) [inline common/events.cpp]: CACHE_DRAIN is set to 1 second, but we want to read for 2 seconds

File: common/events.cpp [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+                    event_data[EVENT_STR_DATA].substr(0, 32).c_str());
+            
+            key = ev.begin()->first;
+
+            rc = deserialize(ev.begin()->second, params);
+            RET_ON_ERR(rc == 0, "failed to deserialize params %s",
+                    ev.begin()->second.substr(0, 32).c_str());
+
+        }
+        else {
```
Comment (zbud-msft) [inline common/events.cpp:384]: Empty else block

File: common/events.cpp [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+
+/* Expect only one subscriber per process */
+static EventSubscriber *s_subscriber = NULL;
+
+event_handle_t
+events_init_subscriber(bool use_cache, int recv_timeout,
+        const event_subscribe_sources_t *sources)
+{
+    if (s_subscriber == NULL) {
+        EventSubscriber *p = new EventSubscriber();
```
Comment (zbud-msft) [inline common/events.cpp]: Wouldn't s or sub be a better name than p?

File: common/events_common.h [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        r += "&";
+    else if (std::is_rvalue_reference<T>::value)
+        r += "&&";
+    return r;
+}
+
+
+template <class T>
+std::string
+get_typename(T &val)
```
Comment (zbud-msft) [inline common/events_common.h:94]: Why is address of value required here?

File: common/events_common.h (hunk 1 above)
Comment (renukamanavalan) [reply common/events_common.h]: Guess, any std:: use is redundant. So removed all std:: uses.

File: common/events_common.h (hunk 2 above)
Comment (renukamanavalan) [reply common/events_common.h:174]: It is a template typename
It could be XYZ too

File: common/events.cpp (hunk 3 above)
Comment (renukamanavalan) [reply common/events.cpp]: Where is this 2 coming from ?
This is just to drain the ZMQ's cache at the receiver end. 
We do this after cache-service is started, which will connect in milliseconds. So reading for a second is much more than required.

File: common/events.cpp (hunk 4 above)
Comment (renukamanavalan) [reply common/events.cpp:384]: Compiler will compile it out. More for code readers, to understand the else clause.

File: common/events.cpp (hunk 5 above)
Comment (renukamanavalan) [reply common/events.cpp]: AGREE!
Single letter naming has costed me heavily before ;)

File: common/events_common.h (hunk 6 above)
Comment (renukamanavalan) [reply common/events_common.h:94]: It is reference. Pass by reference.
By reference is required to assess the type

File: common/events.cpp (hunk 3 above)
Comment (zbud-msft) [reply common/events.cpp]: Line 208, comment mentions to read for 2 seconds.

File: common/events_common.h (hunk 2 above)
Comment (zbud-msft) [reply common/events_common.h:174]: Just to clarify my question, any need for passing by reference here?

Comment (renukamanavalan) [reply common/events_common.h:174]: Reference is same as pointer, but assures that the pointer can't be NULL.
So where we mandatorily expect data, we could use reference and compiler takes care of it.

File: common/events.cpp (hunk 3 above)
Comment (renukamanavalan) [reply common/events.cpp]: Corrected. Thanks

**Review by zbud-msft — APPROVED** [review]

**Pull request merged**